20240618-RCBS-Conversation-building-block.md

```diff
@@ -0,0 +1,94 @@
+# Conversation building block
+
+* Author(s): Loong Dai (@daixiang0)
+* Updated: 2024-06-18
+
+## Overview
+
+This is a proposal for a new building block for Dapr to allow developers to leverage LLM services in a consistent way. Goal is to expose an API that allows developers to ask Dapr to do request.
+
+## Background
+
+Now there are many large language model servers or toolkits, which provides own APIs, like [OpenAI](https://openai.com/), [Hugging Face](https://huggingface.co/), [Kserve](https://kserve.github.io/website/latest/), [OpenVINO](https://docs.openvino.ai/) and so on.
+
+For developers, it is hard to migrate from one to the other due to hardcode and API differences.
+
+For startups and communities, they need to implement the popular APIs as soon as possible, or users maybe give up tring or adopting because of the cost of migration.
+
+This is an area where Dapr can help. We can offer an abstraction layer on those APIs.
+
+## Components
+
+Components in dapr/components-contrib are to be placed in the `conversation` folder and must implement the `Conversation` interface:
+
+```go
+// Conversation offers an interface to perform low-level conversational operations
+type Conversation interface {
+    // Conversate one conversation
+    Conversate(
+        // Context that can be used to cancel the running operation
+        ctx context.Context, 
+        // Name of the model
+        model_name string,
+        // Inputs for the conversation, support multiple input in one time
+        inputs []byte,
+        // Parameters for all custom fields
+        parameters string,
+        // Key of API token
+        key string,
```

Comment:
It is a common use case when instances are rate-limited and developers want to load-balance over different instances (or use them in some priority basis). The feature can be kept out of scope for this proposal, but can be worth considering.
```suggestion
        keys []string,
```

Reply:
Thank you very much, it should be a great feature! Let me take some time to think more about it.

Reply:
I add endpoint here, one key for one endpoint.

Reply:
Update endpoint to endpoints, support rr load balancing for now.